### ISSUE TITLE
More general scripting interface

### ISSUE THREAD

**Issue by jhenin** [Colvars/colvars#273]
Some proposed changes:

###  Allow for several colvarscript objects to coexist
- remove the static accessor colvarscript_obj(); instead the commands can be entirely implemented by methods using the 'this' pointer

The main reason why all the cv_*() should be member functions is that they need access to a number of private members of the colvarscript, colvarmodule and colvar classes. The only other option is to create public interfaces for all of those members. In the current implementation the cv_() functions already are the interface.

### Improve performance by returning generic objects
- the macro-based implementation has started doing that work by abstracting accesses to the 'result' string.

**Comment by jhenin** [comment]
An idea of the work in progress is there: 1b505205173537f00e4564d3f33e22403b377507

**Comment by giacomofiorin** [comment]
@jhenin I just rebased the partial change that I had sitting for some time, breaking it up in ced126b and following commits. It expands and improves a bit the macro-based implementation that you were referring to. `colvarscript_obj()` is used in fewer places, and should be easy to remove if appropriate.

About whether to remove `colvarscript_obj()`: I am still a bit neutral, as I can see advantages (polymorphism) as well as disadvantages (no shared state between the script objects).  The change that I include there uses a single object, but this can be changed.

I named the branch `colvarscript-files` because the organization of the scripting interface was my main concern at the time, but the branch contains in fact most of what you mentioned in this issue.

**Comment by jhenin** [comment]
This branch does not build:

> colvars/src/colvaratoms.h:104:12: error: ‘class colvarproxy’ has no member named ‘updated_masses’; did you mean ‘atoms_masses’?

same with updates_charges, and a possible typo in tcl_interp_.

**Comment by giacomofiorin** [comment]
Will check the code and push a fix to this branch.  If you get to do that before me, comment here so that we don't duplicate.

**Comment by jhenin** [comment]
Please go ahead, I'll wait for that to proceed.

**Comment by giacomofiorin** [comment]
@jhenin There were several errors due to keeping around a long-lived branch, sorry.  I fixed most of them and force-pushed `colvarscript-files`.  Need to look in more detail, but I'm finally freer now and it should take much quicker.

**Comment by giacomofiorin** [comment]
All fixed now. The new code builds, but is currently not invoked and therefore not tested. It's a single commit, to make rebasing faster. See if you wanted to change anything specifically. Within the next couple of weeks I'll make the changes needed to invoke it and begin a PR.

**Comment by jhenin** [comment]
May I rebase and force-push colvarscript-files, or do you have changes in there?

**Comment by giacomofiorin** [comment]
I don't.

**Comment by jhenin** [comment]
Force pushed with a small additional commit to allow for building VMD. I kept it separate for readability, but it should be quashed with the rest on the next update.

**Comment by giacomofiorin** [comment]
Sounds good.

**Comment by giacomofiorin** [comment]
FYI I just rebased `colvarscript-files` onto `master`, so that the build system for tests is available.

**Comment by jhenin** [comment]
This is taken care of by #316

**Comment by giacomofiorin** [comment]
I disagree that #316 implements the third item.  I have yet to add the functions to convert arguments and result from/to object other than strings.  Planned to do that as a separate PR after #316 is merged.

**Comment by jhenin** [comment]
Fair enough. To be more precise, I thought that given the progress in that direction, it was not so useful to keeping this issue around.
